### ISSUE TITLE
chpldoc generation error due to unrelated block comment before function declaration

### ISSUE THREAD

**Issue by BryantLam** [chapel-lang/chapel#14250]
Bug. chpldoc fails to generate documentation due to misidentification of the commented-out block of code as a doc-comment preceding a function definition.

```chapel
/*
extern {
    static inline int my_c_func() {
        return 42;
    }
}
*/

proc myFunc() {}
```
```terminal
Running Sphinx v1.3.3
loading pickled environment... not yet created
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 2 source files that are out of date
updating environment: 2 added, 0 changed, 0 removed
reading sources... [100%] modules/docbreak                                                                                                                                                                        

Warning, treated as error:
/tmp/chpldoc-user-2943.deleteme/source/modules/docbreak.rst:19: WARNING: Definition list ends without a blank line; unexpected unindent.

error: building html output from chpldoc sphinx project
```

Related #14207. Perhaps chpldoc should have chosen a different default comment style such as Javadoc `/**` or Doxygen `///`.

**Comment by lydia-duncan** [comment]
We did also consider making it so that documentation comments had to be defined immediately before the symbol being documented, but that seemed too restrictive to users at the time.  That said, there's a fuzzy line where it starts being more and more likely that a comment block is not intended as a documentation comment given its distance from the symbol itself.  Thoughts on that strategy?

Cases like this are also why we added the ability to redefine the documentation comment style, though it sounds like the one you prefer is currently unusable . . .

**Comment by BryantLam** [comment]
> We did also consider making it so that documentation comments had to be defined immediately before the symbol being documented

I'm okay with this approach. It was more surprising to me that `chpldoc` allowed whitespace/newlines to break the comment from the declaration.


> Cases like this are also why we added the ability to redefine the documentation comment style, though it sounds like the one you prefer is currently unusable . . .

Yeah... This is more of a lesson-learned + precedence styling choice. Javadoc uses `/**` for doc comments, which Doxygen inherited for C/C++, but they also added `///` as the default comment style to avoid the non-nesting behavior of block comments.

**Comment by lydia-duncan** [comment]
In my notes from when we added chpldoc, it says that other documentation tools do allow multiple lines of whitespace between the comment and the symbol, though I must admit that I've never intentionally used more than one.

**Comment by BryantLam** [comment]
I'm okay with a less restrictive approach to whitespace if the default comment style was changed (#14251). It's potentially error-prone to allow multiple lines of whitespace when doc comments and normal block comments are the same.

Like in this case, It also makes it less apparent how the user would "opt out" of the doc comment without knowing to use `pragma "no doc".

I can't say I care for the excess whitespace though. :smile:

**Comment by mppf** [comment]
I think it'd be reasonable to not identify it as a doc comment if there are blank lines in between.

I don't think we should have a special docs-comment introduction `/**` etc; see https://github.com/chapel-lang/chapel/issues/14251#issuecomment-540561023

**Comment by lydia-duncan** [comment]
> Like in this case, It also makes it less apparent how the user would "opt out" of the doc comment without knowing to use `pragma "no doc".

Note that you can prevent it by adding an intervening comment:

```chapel
/* This long comment shouldn't get attached to foo, because there is a single
   line comment in the way */

// Blocking!
proc foo() {

}
```

This results in `foo` having documentation generated for it, but no comment associated with it (so all that is documented is the signature)

**Comment by mppf** [comment]
But note that in #14251 we (or maybe just me?) are considering making `//` comments be docs comments.

I think we need some user-facing way to do `pragma "no doc"` either way, though, and I would lean towards that being the actual solution here (possibly combined with different handling of blank lines). Issue #14141 proposes an attribute syntax, which we could use for this. For `no doc` specifically though, I think it would also be reasonable to write a docs comment disabling documentation generation, e.g. `/* nodoc */`. This approach is discussed in issue #13115.

**Comment by lydia-duncan** [comment]
> But note that in #14251 we (or maybe just me?) are considering making // comments be docs comments.

The point I was trying to make was not aimed at specific comment style, the point was that *any* intervening blank comment will avoid including comments that were not intended, without preventing the symbol itself from being present in the documentation (if that was desired).  I could just as easily have written `/* */` on that line and the long comment would have been avoided.

#13115 also discusses alternatives to `pragma "no doc"`

**Comment by BryantLam** [comment]
> the point was that any intervening blank comment will avoid including comments that were not intended, without preventing the symbol itself from being present in the documentation (if that was desired)

I don't like this behavior. As a user, I'm expected to now remember how to opt-out of documenting when using my comments and the associated tricks of how to not document something vs. choosing when to explicitly document.

Also, the idea of `/* nodoc */` also means that comments have more special behaviors. I'd prefer an attribute or something else.

**Comment by mppf** [comment]
> I don't like this behavior. 

These things are subjective, and I can't really argue with an opinion. But...

> As a user, I'm expected to now remember how to opt-out of documenting when using my comments and the associated tricks of how to not document something vs. choosing when to explicitly document.

Isn't this also true if we have say doxygen-style docs comments `/**` only? Maybe what you are trying to say is that you think it's easier to remember to use a special form for docs comments than it is to remember something like "the last comment before it is the docs comment"?

I think you are saying that there is something weird/surprising about having

``` chapel
/* */
proc foo() { }
```

and then having the generated docs list `proc foo` with no description.

But how can that be surprising when the doc-comment string is empty? I feel I am missing something about your argument (unless it is simply about `/**` vs `/*` for introducing docs comment, but we should try to keep discussion about that in #14251).

**Comment by BryantLam** [comment]
> Maybe what you are trying to say is that you think it's easier to remember to use a special form for docs comments than it is to remember something like "the last comment before it is the docs comment"?

Yes. The rest of your comment is a corollary to this statement. I guess what I'm saying is that it's harder to make a mistake -- or in this case -- forget about when a comment turns into a document comment if it is explicitly done by the user (`/**` or `///` or `//!`, etc.).

**Comment by lydia-duncan** [comment]
> As a user, I'm expected to now remember how to opt-out of documenting when using my comments and the associated tricks of how to not document something vs. choosing when to explicitly document.

I mean, the main expected action for a user is still "comment your functions properly".  It's a pretty minimal penalty for something they should be doing anyways, and the current situation rewards users that comment reasonably by habit.

**Comment by mppf** [comment]
Or, maybe the thing to do here is for `chpldoc` to warn if the comment it found is not on the preceeding line (i.e. there is a blank line in-between).
